lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -66,116 +63,56 @@ module Cfg = Graphlib.Make(Node)(Edge)
 
 type cfg = Cfg.t [@@deriving compare]
 
+module Visited = struct
 
-(** Interval tree spanning visited memory *)
-module type Span = sig
-  type range = addr * addr
-  type t [@@deriving sexp_of]
+  type entry = Unknown | Insn of {last : addr}
+  type t = entry Addr.Map.t
 
-  val empty : t
-  val add : t -> range -> t
-  val mem : t -> addr -> bool
-  val upper_bound : t -> addr -> addr option
-  val min : t -> addr option
-  val max : t -> addr option
-  val pp : Format.formatter -> t -> unit
-end
+  let empty = Addr.Map.empty
+  let add t mem =
+    Map.set t (Memory.min_addr mem) (Insn {last = Memory.max_addr mem})
 
-module Span : Span = struct
-  module Range = struct
-    module T = struct
-      type t = addr * addr [@@deriving sexp]
-      let compare (a1,_) (a2,_) = Addr.compare a1 a2
-    end
-    type t = T.t [@@deriving sexp]
-    include Comparable.Make(T)
-  end
-
-  type range = Range.t
-
-  let sexp_of_range (a0,a1) =
-    Sexp.List [
-      Sexp.Atom (Addr.string_of_value a0);
-      Sexp.Atom (Addr.string_of_value a1);
-    ]
-
-  type t =
-    | Empty
-    | Node of t * range * addr * t
-  [@@deriving sexp_of]
-
-  let rec pp fmt = function
-    | Empty -> ()
-    | Node (lhs,(x,y),_,rhs) ->
-      Format.fprintf fmt "@[%a(%a,%a)@,%a@]"
-        pp lhs Addr.pp x Addr.pp y pp rhs
-
-  let empty = Empty
-
-  let rng_max (_,a1) a2 = Addr.max a1 a2
-
-  (** @pre: x <= p *)
-  let intersects (x,y) (p,q) = p <= y
-  (** [merge x y] if [intersects x y]  *)
-  let merge (x,p) (_,q) : range = (x, Addr.max p q)
-
-  let rec add t ((_,a1) as rng) = match t with
-    | Empty -> Node (Empty,rng,a1,Empty)
-    | Node (lhs,top,m,rhs) ->
-      let m = rng_max rng m in
-      if Range.(rng < top) then
-        if intersects rng top
-        then Node (lhs, merge rng top, m, rhs)
-        else Node (add lhs rng, top, m, rhs)
-      else if intersects top rng
-      then Node (lhs, merge top rng, m, rhs)
-      else Node (lhs, top, m, add rhs rng)
-
-  let rec mem t addr = match t with
-    | Empty -> false
-    | Node (lhs,(a0,a1),mx,rhs) ->
-      if Addr.(addr > mx) then false
-      else if Addr.(addr < a0) then mem lhs addr
-      else if Addr.(addr >= a1) then mem rhs addr
-      else true
-
-  (** [upper_bound visited addr] returns minimum address that is
-      greater than [addr] and is visited. Returns [None] if no such
-      address was found *)
-  (* returns the lowest left bound greater than [addr] *)
-  let rec upper_bound t a = match t with
-    | Empty -> None
-    | Node (_,_,mx,_) when Addr.(a > mx) -> None
-    | Node (lhs,(a0,_),_,rhs) ->
-      if Addr.(a > a0) then upper_bound rhs a
-      else match upper_bound lhs a with
-        | None -> Some a0
-        | some -> some
-
-  let rec min = function
-    | Empty -> None
-    | Node (Empty,(a0,_),_,_) -> Some a0
-    | Node (lhs,_,_,_) -> min lhs
-
-  let max = function
-    | Empty -> None
-    | Node (_,_,m,_) -> Some m
-end
+  let visit t addr =
```

Comment:
the name `touch` would be better, but this function has weird semantics. It behaves in the DWIM manner, and eventually, when `I` or `M` will change, it will bring lots of surprises. The right solution is to be as explicit as possible, and have two functions, named `add_insn` and `add_unknown` or parameterize `add` with the `entry` type. And the call explicitly in all handlers.